File: frontend/src/api/configs.ts
import apiClient from './client';
import type { Node, Edge } from '@xyflow/react';
import type { DeviceNodeData } from '../types';

// ─── リクエスト / レスポンス型 ─────────────────────────────────────────────────

export interface InterfaceConfig {
  name: string;
  slot: number;
  ip_address: string;
  prefix_length: number;
  vlan_id: string;
  vrf_name: string;
}

export interface NodeConfigRequest {
  node_id: string;
  label: string;
  vendor: string;
  os: string;
  node_definition: string;
  x: number;
  y: number;
  interfaces: InterfaceConfig[];
  extra_params: Record<string, unknown>;
}

export interface NodeConfigResponse {
  node_id: string;
  label: string;
  config: string;
}

export interface TopologyConfigResponse {
  configs: NodeConfigResponse[];
}

export interface EdgeData {
  src_node_id: string;
  src_iface_slot: number;
  dst_node_id: string;
  dst_iface_slot: number;
}

export interface CMLPushRequest {
  title: string;
  description: string;
  nodes: NodeConfigRequest[];
  edges: EdgeData[];
  start_after_push: boolean;
}

export interface CMLLabResponse {
  lab_id: string;
  title: string;
  state: string | null;
}

// ─── 変換ヘルパー ────────────────────────────────────────────────────────────

/** React Flow ノードを API リクエスト形式に変換する */
export function nodeToConfigRequest(node: Node<DeviceNodeData>): NodeConfigRequest {
  const { template, interfaces, label } = node.data;
  return {
    node_id: node.id,
    label,
    vendor: template.vendor,
    os: template.os,
    node_definition: template.node_definition,
    x: Math.round(node.position.x),
    y: Math.round(node.position.y),
    interfaces: template.interfaces
      .filter((i) => i.type !== 'management')
      .map((i, index) => ({
        name: i.name,
        slot: i.slot ?? index,
        ip_address: interfaces[i.name]?.ip_address ?? '',
        prefix_length: interfaces[i.name]?.prefix_length ?? 24,
        vlan_id: interfaces[i.name]?.vlan_id ?? '',
        vrf_name: interfaces[i.name]?.vrf_name ?? '',
      })),
    extra_params: {},
  };
}

/** React Flow エッジを API リクエスト形式に変換する */
export function edgeToEdgeData(edge: Edge): EdgeData {
  return {
    src_node_id: edge.source,
    src_iface_slot: parseInt(edge.sourceHandle ?? '0', 10),
    dst_node_id: edge.target,
    dst_iface_slot: parseInt(edge.targetHandle ?? '0', 10),
  };
}

// ─── API 呼び出し ─────────────────────────────────────────────────────────────

export async function generateConfigs(
  nodes: NodeConfigRequest[],
): Promise<TopologyConfigResponse> {
  const { data } = await apiClient.post<TopologyConfigResponse>('/configs/preview', { nodes });
  return data;
}

export async function pushToCML(request: CMLPushRequest): Promise<CMLLabResponse> {
  const { data } = await apiClient.post<CMLLabResponse>('/configs/push-to-cml', request);
  return data;
}

File: backend/app/routers/configs.py
import time

import yaml
from fastapi import APIRouter, HTTPException

from app.schemas.config import (
    CMLLabResponse,
    CMLPushRequest,
    NodeConfigResponse,
    TopologyConfigRequest,
    TopologyConfigResponse,
)
from app.services.cml_client import cml_client
from app.services.config_generator import config_generator

router = APIRouter(prefix="/configs", tags=["configs"])


@router.post("/preview", response_model=TopologyConfigResponse)
def preview_configs(data: TopologyConfigRequest) -> TopologyConfigResponse:
    """各ノードの Config をプレビュー生成して返す（CML への送信なし）"""
    configs: list[NodeConfigResponse] = []
    for node in data.nodes:
        try:
            config_text = config_generator.generate(node)
        except FileNotFoundError:
            config_text = f"# No template found for {node.vendor}/{node.os}\n"
        except Exception as exc:
            config_text = f"# Config generation error: {exc}\n"
        configs.append(
            NodeConfigResponse(node_id=node.node_id, label=node.label, config=config_text)
        )
    return TopologyConfigResponse(configs=configs)


@router.post("/push-to-cml", response_model=CMLLabResponse)
def push_to_cml(data: CMLPushRequest) -> CMLLabResponse:
    """Config を生成し CML へトポロジをインポートする"""
    # 1. 各ノードの Config を生成
    node_configs: dict[str, str] = {}
    for node in data.nodes:
        try:
            node_configs[node.node_id] = config_generator.generate(node)
        except Exception:
            node_configs[node.node_id] = ""

    # 2. React Flow ID → CML node ID のマッピングを作成
    node_id_map: dict[str, str] = {}
    cml_nodes = []
    for i, node in enumerate(data.nodes):
        cml_id = f"n{i}"
        node_id_map[node.node_id] = cml_id

        interfaces_yaml = [
            {"id": f"i{iface.slot}", "label": iface.name, "slot": iface.slot, "type": "physical"}
            for iface in node.interfaces
        ]

        cml_nodes.append(
            {
                "id": cml_id,
                "label": node.label,
                "node_definition": node.node_definition,
                "x": node.x,
                "y": node.y,
                "configuration": node_configs.get(node.node_id) or None,
                "interfaces": interfaces_yaml,
                "ram": None,
                "cpus": None,
                "cpu_limit": None,
                "boot_disk_size": None,
                "data_volume": None,
                "hide_links": False,
                "image_definition": None,
                "tags": [],
            }
        )

    # 3. リンクを構築
    links = []
    for i, edge in enumerate(data.edges):
        src = node_id_map.get(edge.src_node_id)
        dst = node_id_map.get(edge.dst_node_id)
        if not src or not dst:
            continue
        links.append(
            {
                "id": f"l{i}",
                "i1": f"{src}[i{edge.src_iface_slot}]",
                "i2": f"{dst}[i{edge.dst_iface_slot}]",
            }
        )

    # 4. CML topology YAML を構築
    topology = {
        "lab": {
            "title": data.title,
            "description": data.description,
            "notes": "",
            "timestamp": time.time(),
            "version": "0.1.0",
        },
        "nodes": cml_nodes,
        "links": links,
    }
    topology_yaml = yaml.dump(topology, allow_unicode=True, sort_keys=False)

    # 5. CML へインポート
    try:
        result = cml_client.import_topology(topology_yaml)
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"CML connection error: {exc}") from exc

    return CMLLabResponse(
        lab_id=result["id"],
        title=result["title"],
        state=result.get("state"),
    )
